app/agents/JobOfferManagerAgent.py

```diff
@@ -93,32 +97,49 @@ class AwaitApplication(spade.behaviour.PeriodicBehaviour):
     agent: JobOfferManagerAgent
 
     async def run(self):
-        self.agent.logger.info("AwaitApplication behaviour run.")
-        new_applications = self.check_new_applications()
-        if len(new_applications) > 0:
-            await self.process_candidates(new_applications)
-
-        processed_applications = self.agent.jobOfferModule.get_processed_applications(
-            self.agent.job_offer_id
-        )
-        if len(processed_applications) > 0:
-            await self.trigger_analysis(processed_applications)
-
-    def check_new_applications(self) -> list[ApplicationDetails]:
+        if self.agent.jobOffer.status != JobOfferStatus.CLOSED:
+            self.agent.logger.info("AwaitApplication behaviour run.")
+            applications_to_process = self.check_applications_to_process()
+            if len(applications_to_process) > 0:
+                await self.process_candidates(applications_to_process)
+                await self.trigger_analysis(applications_to_process)
+        self.agent.jobOffer = self.agent.jobOfferModule.get(self.agent.job_offer_id)
+
+    def check_applications_to_process(self) -> list[ApplicationDetails]:
         new_applications = self.agent.jobOfferModule.get_new_applications(
             self.agent.job_offer_id
         )
 
-        # ProcessApplication activity
+        applications_to_process = []
+        applications_to_reject = []
+        old_applications = len(self.agent.jobOffer.applications)
+
+        for app in new_applications:
+            if old_applications + 1 <= self.agent.jobOffer.max_number_of_candidates:
+                applications_to_process.append(app)
+                old_applications = old_applications+1
+            else:
+                applications_to_reject.append(app)
+
         result = self.agent.jobOfferModule.change_application_status(
             self.agent.job_offer_id,
-            [x.candidate_id for x in new_applications],
+            [x.candidate_id for x in applications_to_process],
             ApplicationStatus.PROCESSED,
-        )
+            )   
+
+        _ = self.agent.jobOfferModule.change_application_status(
+            self.agent.job_offer_id,
+            [x.candidate_id for x in applications_to_reject],
+            ApplicationStatus.REJECTED,
+            )
```

Comment:
1. lepszym rozwiązaniem wydaje mi się że by było czekanie na `max_number_of_candidates` pierwszych zakończonych rekrutacji a nie po prostu wpłynięć aplikacji, ale to po prostu inna logika, nie ma znaczenia
2. na poczatku myslalem ze `REJECTED` to odrzucona w sensie że nie został wybrany jako best candidate - w jakim statusie wtedy będzie aplikacja? finished?

Reply:
1. Zdalem sobie sprawe z tego jak juz to zaimplmentowalem. Najwyzej sie poprawi w przyszlosci. Nie beda to wielkie zmiany. 
2. Dokładnie tak. finished ale nie w best candidate. W przyszlosci mozna to poprawic na wiele najlepszych kandydatow i inna logike traktowania aplikacji. 

Na ten moment zamykam bo nie mam czasu wiecej pracowac nad tym, ale za jakis czas moge to poprawic.